### PR TITLE
Updated e0072 bonus to new format

### PR DESCRIPTION
**Pull request by phungleson**
Will do this in several steps

- [ ] Refactor the code to use `name: Spanned<Name>`
- [ ] Calculate name.span 
- [ ] Display name.span correctly in error message

### CODE REVIEW

File: src/librustc/hir/fold.rs [hunk 1]
```diff
@@ -850,7 +850,7 @@ pub fn noop_fold_crate<T: Folder>(Crate { module, attrs, config, span,
     let config = folder.fold_meta_items(config);
 
     let crate_mod = folder.fold_item(hir::Item {
-        name: keywords::Invalid.name(),
+        name: spanned(BytePos(0), BytePos(0), keywords::Invalid.name()),
```
Comment (KiChjang) [inline src/librustc/hir/fold.rs]: There should a DUMMY_SP in libsyntax_pos that you can use for names that don't have a corresponding span. Don't use it for the parser though.

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -752,7 +753,7 @@ impl<'a> LoweringContext<'a> {
 
         hir::Item {
             id: i.id,
-            name: i.ident.name,
+            name: spanned(BytePos(0), BytePos(0), i.ident.name),
```
Comment (KiChjang) [inline src/librustc/hir/lowering.rs]: I'm kinda confused, why isn't this `i.ident.name` not a `Spanned<Name>`?

File: src/librustc/hir/fold.rs [hunk 3]
```diff
@@ -894,7 +894,7 @@ pub fn noop_fold_item<T: Folder>(item: Item, folder: &mut T) -> Item {
 
     Item {
         id: id,
-        name: folder.fold_name(name),
+        name: spanned(BytePos(0), BytePos(0), folder.fold_name(name.node)),
```
Comment (KiChjang) [inline src/librustc/hir/fold.rs]: Same as above.

File: src/librustc/hir/lowering.rs (hunk 2 above)
Comment (KiChjang) [inline src/librustc/hir/lowering.rs]: Actually, hold on. HIR types shouldn't have a span, only AST ones, so you should revert changes that you've made in hir/mod.rs.

Comment (phungleson) [inline src/librustc/hir/lowering.rs]: Oh ok sure, so we don't need to add `span` to `item.name` in hir.

Could you give some idea about the other direction, so I can close this PR and make the changes a new one?

Comment (phungleson) [inline src/librustc/hir/lowering.rs]: Actually I have a question about the differences between `Item` in `ast` and one in `hir`?

Comment (KiChjang) [inline src/librustc/hir/lowering.rs]: AST -> HIR only goes in one direction, there isn't a mechanism to change an HIR node back into an AST node.

Comment (KiChjang) [inline src/librustc/hir/lowering.rs]: Treat HIR nodes as simplified versions of its AST counterpart, they're mostly the same but HIR nodes are desugared and have less information on them.

**Comment by phungleson** [comment]
Thanks @KiChjang let me close this one an open another PR in rust to make this more official

**Pull request closed without merging**